metrics-operator/controllers/metrics/controller.go

```diff
@@ -101,23 +101,56 @@ func (r *KeptnMetricReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 	reconcile := ctrl.Result{Requeue: true, RequeueAfter: 10 * time.Second}
 
-	value, rawValue, err := r.getResults(ctx, metric, provider, metricProvider)
-	if err != nil {
-		r.Log.Error(err, "Failed to evaluate the query", "Response from provider was:", (string)(rawValue))
-		metric.Status.ErrMsg = err.Error()
-		metric.Status.Value = ""
-		metric.Status.RawValue = cupSize(rawValue)
-		metric.Status.LastUpdated = metav1.Time{Time: time.Now()}
-		reconcile = ctrl.Result{Requeue: false}
+	if metric.Spec.Range.StoredResults > 0 {
+		value, _, err := r.getResults(ctx, metric, provider, metricProvider)
+		if err != nil {
+			intervalResult := metricsapi.IntervalResult{
+				Value:       "",
+				Range:       nil,
+				LastUpdated: metav1.Time{Time: time.Now()},
+				ErrMsg:      err.Error(),
+			}
+			if len(metric.Status.IntervalResults) >= int(metric.Spec.Range.StoredResults) {
+				metric.Status.IntervalResults = append(metric.Status.IntervalResults[1:], intervalResult)
+			} else {
+				metric.Status.IntervalResults = append(metric.Status.IntervalResults, intervalResult)
+			}
+		} else {
+			intervalResult := metricsapi.IntervalResult{
+				Value:       value,
+				Range:       metric.Spec.Range,
+				LastUpdated: metav1.Time{Time: time.Now()},
+				ErrMsg:      err.Error(),
+			}
+			if len(metric.Status.IntervalResults) >= int(metric.Spec.Range.StoredResults) {
+				metric.Status.IntervalResults = append(metric.Status.IntervalResults[1:], intervalResult)
+			} else {
+				metric.Status.IntervalResults = append(metric.Status.IntervalResults, intervalResult)
+			}
+		}
+		if err := r.Client.Status().Update(ctx, metric); err != nil {
```

Comment:
the update can be done at the end of the function after the `if`, so we don't have to duplicate it for both branches of the if statement